### PR TITLE
Fix checkboxes only shown on crengine; local retval

### PR DESCRIPTION
**Pull request by zwim**
This PR fix #7883  
Do not show checkboxes on pdf and mupdf documents, don't crash.

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/7947)

### CODE REVIEW

**Comment by zwim** [comment]
@poire-z 
https://github.com/koreader/koreader/pull/7883#discussion_r658030644

If I use `self.ui.document.info.has_pages` the check buttons are shown, with `self.ui.document.provider == "crengine"` they are not shown.

File: frontend/apps/reader/modules/readersearch.lua [hunk 1]
```diff
@@ -177,19 +177,21 @@ function ReaderSearch:onShowFulltextSearchInput()
...
-    local checkbox_shift = math.floor((self.input_dialog.width - self.input_dialog._input_widget.width) / 2 + 0.5)
-    local check_buttons = HorizontalGroup:new{
-        HorizontalSpan:new{width = checkbox_shift},
-        VerticalGroup:new{
-            align = "left",
-            self.check_button_case,
-            self.check_button_regex,
-        },
-    }
+    if self.ui.document.provider == "crengine" then
```
Comment (poire-z) [inline frontend/apps/reader/modules/readersearch.lua:180]: Noooo, I don't want to see any more of that ! :)
https://github.com/koreader/koreader/pull/7883#discussion_r658030644

**Review by poire-z — COMMENTED** [review]
Sure (or can you check quickly, even if you don't enable it with this PR) that case sensitive seach can't be made to work with PDFs?

**Comment by poire-z** [comment]
> If I use `self.ui.document.info.has_pages` the check buttons are shown, with 

So.... use `not self.ui.document.info.has_pages`  :) (PDF has_pages, EPUBs don't)

**Comment by zwim** [comment]
Case sensitive seems to work on an epub opend with mupdf. :)

**Comment by zwim** [comment]
One thing is strange. Full text search works on a pdf (Micro typography thesis), epubs (mupdf and crengine) but not on the `2col.pdf` test document. Is that OK`

File: frontend/apps/reader/modules/readersearch.lua [hunk 2]
```diff
@@ -177,22 +177,31 @@ function ReaderSearch:onShowFulltextSearchInput()
...
+                self.check_button_case,
+            },
+        }
+    else
+        check_buttons = HorizontalGroup:new{
             HorizontalSpan:new{width = checkbox_shift},
             VerticalGroup:new{
                 align = "left",
                 self.check_button_case,
                 self.check_button_regex,
```
Comment (poire-z) [inline frontend/apps/reader/modules/readersearch.lua]: You can avoid this duplication with (luckily, it is last so we can make it nil):
```
        check_buttons = HorizontalGroup:new{
            HorizontalSpan:new{width = checkbox_shift},
            VerticalGroup:new{
                align = "left",
                self.check_button_case,
                not self.ui.document.info.has_pages and self.check_button_regex or nil,
```
(not sure if the `or nil` is needed and if having it `false` is ok).

Comment (zwim) [reply frontend/apps/reader/modules/readersearch.lua]: `or nil` is necessary.

**Comment by poire-z** [comment]
> One thing is strange. Full text search works on a pdf (Micro typography thesis), epubs (mupdf and crengine) but not on the 2col.pdf test document. Is that OK?

I guess that's ok. Some PDF may have usable text layers, others not. @Frenzie will know more :)

**Comment by Frenzie** [comment]
@zwim 
> One thing is strange. Full text search works on a pdf (Micro typography thesis), epubs (mupdf and crengine) but not on the 2col.pdf test document. Is that OK?

That depends on what you're asking. The answer's yes, because you didn't break it, and yes, because the document doesn't have a text layer.

Now if your question were whether it's okay that you broke it, that's not as easy to say. ;-)

**Comment by zwim** [comment]
What do you mean with **it**? `2col.pdf` or the other PR?

If you mean the later one am really not happy to break the full text search on non crengine document :((
But that problem arose as I dropped the regex-search menu (which was good and did not have that problem). Of course the version with the check buttons is more user-friendly and more professional. I like it.

But after all we win a case-sensitive search on all documents :)
`+` + `-` = `+` at the end of the day ;)

**Comment by zwim** [comment]
@Poire-z: I noticed circleci is working now!!!! :100: :100:

I got an eMail from circleci, about that. So things will get better on this front.

**Comment by poire-z** [comment]
> Now if your question were whether it's okay that you broke it, that's not as easy to say. ;-)

I guess he means breaking 2col.pdf. But no, you didn't broke it: full text search already did not work with it with v2021.05.

**Comment by poire-z** [comment]
(And full text search is and was broken/strange on PDF. Search for "Adapt" case sensitive or not in our paper.pdf, and lots of times, you switch pages but not hit is highlighten.)

**Comment by Frenzie** [comment]
"Were" is the subjunctive. In other words, a hypothetical describing a situation that hasn't occurred in reality. The answer to your question is that yes, it's okay if 2col.pdf (and very specifically 2col.pdf) is "broken" because it's supposed to be. It doesn't have a text layer.

**Comment by poire-z** [comment]
Long-press on `(hold for help)` makes and leaves the input dialog translucent.

I hope we won't have too many issues about how to regex search :) - took me a few tries (thanks for keeping the previous input!) to come up with this to match France but not François, without getting fulle sentences:
image

**Comment by poire-z** [comment]
> Long-press on (hold for help) makes and leaves the input dialog translucent.

Not sure how to fix that, but I'll merge this later anyway even if not fixed.
We have these in button.lua:
https://github.com/koreader/koreader/blob/826a7657057132069c0b1ec122b6765ab77646d6/frontend/ui/widget/button.lua#L139-L145
https://github.com/koreader/koreader/blob/826a7657057132069c0b1ec122b6765ab77646d6/frontend/ui/widget/button.lua#L397-L425
May be CheckButton needs them too.

**Comment by zwim** [comment]
> Long-press on `(hold for help)` makes and leaves the input dialog translucent.

I can confirm that. But I am sure that is new quirk (I have tested it a lot of time during implentation, and I did not notice that.) 


> I hope we won't have too many issues about how to regex search :) -

->> What about a wiki entry: `Sesame street` or how to find a word starting with ...

> took me a few tries (thanks for keeping the previous input!) to come up with this to match France but not François, without getting fulle sentences:

I think the Switzerland trip mad me really tired. I is the point that you got what you want (but it was difficult)?

**Comment by zwim** [comment]
> Long-press on (hold for help) makes and leaves the input dialog translucent.

Interestingly, if you move the infomessage the input dialog does not get translucent.

**Comment by zwim** [comment]
And if you move the input dialog before long press, long press for help, the dialog is also not tranlucent.

**Comment by poire-z** [comment]
> I can confirm that. But I am sure that is new quirk (I have tested it a lot of time during implentation, and I did not notice that.)

Possibly, You started the PR so long ago that in the meantime, @hius07 made InputDialog movable and transclucent :) So, you got that by rebasing.
We'll fix that later, may be @hius07 will have a simpler idea (too late for me to :) but I can make CheckButton look alike Button if that's the real fix).

> And if you move the input dialog before long press for help, the dialog is also not translucent.
Maybe this is just and initialization problem?

No, I think that's a feature of MovableContainer: if you have moved it, the first long-press moves it back to its initial position.

> Is the point that you got what you want (but it was difficult)?

Yes, I got what I wanted. Just thinking that most people would want to lookup simple words, so a 3rd chekbox` [ ] Find minimal/words` that changes all `*` to `*?` and prepend and append `\b` could be useful :) (Just a stupid idea, in case we find another less stupid idea :)

**Comment by hius07** [comment]
After reading https://github.com/koreader/koreader/pull/3636 once more, have a general question:
Do we need transparency for input dialogs at all? Is it useful?

**Comment by poire-z** [comment]
> Do we need transparency for input dialogs at all? Is it useful?

It's the generic behaviour of MovableContainer. If you'd want to disable transparency, you'd need to hack it and make it even more complex. (It's probably not useful, but it's also not really useful on many others ;) and well, I don't need to teach you Consistency :)

But it would not fix the bug with these CheckButtons: move this fulltext search dialog, long-press on the checkbutton: the dialog is moved back to its original position (which is the same kind of but that we have to solve :)

**Comment by hius07** [comment]
> you'd need to hack it and make it even more complex.

Can we use `alpha = 1` passed from the input dialog to the MovableContainer?


> long-press on the checkbutton: the dialog is moved back to its original position

Oh yes, that is unexpected by a user.

**Comment by hius07** [comment]
Anyway, your `Button:onHoldReleaseSelectButton() ` is perfect.

**Comment by poire-z** [comment]
> Can we use alpha = 1 passed from the input dialog to the MovableContainer?

Oh, I guess that would work. But then, being familiar with movable stuff, I would be surprised (and would keep long-pressing) if its behaviour was different than others. And if it's not useful to me to get it translucent, well, I would just not long-press :)

> Anyway, your Button:onHoldReleaseSelectButton() is perfect.

Just looking at the code ? or have you tested/done it / want to do it ? Or should I?

**Comment by hius07** [comment]
> or have you tested

For buttons: upper row of buttons in SkimToWidget.
No, I haven't tested for checkbuttons.

**Comment by hius07** [comment]
Are these variables used anywhere?
https://github.com/koreader/koreader/blob/86a1552ccd7d691fcd862b27db09b17f87e9ef4c/frontend/apps/reader/modules/readersearch.lua#L90-L91

**Comment by zwim** [comment]
@hius07: You are right, the variables are a leftover. I have a PR almost ready for saving search settings. I will remove them in there. Thanks.

**Pull request merged**